File: src/services/fraudPreventionService.ts
/**
 * Service de prévention de fraude pour l'inscription
 * Version Web - Identique au mobile Flutter
 * 
 * Fonctionnalités :
 * - Device fingerprinting (navigateur)
 * - Détection IP
 * - Vérification blacklist
 * - Score de fraude (0-100)
 * - Logging des tentatives
 */

import { supabase } from '../lib/supabase';

export interface FraudCheckResult {
  fraudScore: number; // 0-100
  isSuspicious: boolean;
  flags: FraudFlag[];
  recommendation: 'allow' | 'manual_review' | 'block';
}

export interface FraudFlag {
  type: string;
  severity: 'low' | 'medium' | 'high' | 'critical';
  message?: string;
}

export interface SignupAttempt {
  email: string;
  phone?: string;
  deviceFingerprint: string;
  ipAddress: string;
  userAgent: string;
  stepReached: number;
  success: boolean;
  failureReason?: string;
}

class FraudPreventionService {
  private deviceFingerprint: string | null = null;
  private userIpAddress: string | null = null;

  /**
   * Génère une empreinte unique du navigateur
   */
  async generateDeviceFingerprint(): Promise<string> {
    if (this.deviceFingerprint) {
      return this.deviceFingerprint;
    }

    const components: string[] = [];

    // User Agent
    components.push(navigator.userAgent);

    // Langue
    components.push(navigator.language);

    // Fuseau horaire
    components.push(String(new Date().getTimezoneOffset()));

    // Résolution écran
    components.push(`${screen.width}x${screen.height}x${screen.colorDepth}`);

    // Plugins (si disponible)
    if (navigator.plugins && navigator.plugins.length > 0) {
      const plugins = Array.from(navigator.plugins)
        .map(p => p.name)
        .sort()
        .join(',');
      components.push(plugins);
    }

    // Canvas fingerprint
    try {
      const canvas = document.createElement('canvas');
      const ctx = canvas.getContext('2d');
      if (ctx) {
        ctx.textBaseline = 'top';
        ctx.font = '14px Arial';
        ctx.fillStyle = '#f00';
        ctx.fillRect(0, 0, 100, 50);
        ctx.fillStyle = '#069';
        ctx.fillText('ChecksFleet', 2, 15);
        components.push(canvas.toDataURL());
      }
    } catch (e) {
      // Canvas peut être bloqué par la politique de sécurité
      components.push('canvas-blocked');
    }

    // WebGL fingerprint
    try {
      const gl = document.createElement('canvas').getContext('webgl');
      if (gl) {
        const debugInfo = gl.getExtension('WEBGL_debug_renderer_info');
        if (debugInfo) {
          components.push(gl.getParameter(debugInfo.UNMASKED_VENDOR_WEBGL));
          components.push(gl.getParameter(debugInfo.UNMASKED_RENDERER_WEBGL));
        }
      }
    } catch (e) {
      components.push('webgl-blocked');
    }

    // Platform
    components.push(navigator.platform);

    // Do Not Track
    components.push(String(navigator.doNotTrack));

    // Hardware concurrency (nombre de CPU)
    if (navigator.hardwareConcurrency) {
      components.push(String(navigator.hardwareConcurrency));
    }

    // Device memory (si disponible)
    if ('deviceMemory' in navigator) {
      components.push(String((navigator as any).deviceMemory));
    }

    // Générer hash simple
    const fingerprint = await this.simpleHash(components.join('|'));
    this.deviceFingerprint = fingerprint;
    
    return fingerprint;
  }

  /**
   * Hash simple (non cryptographique mais suffisant pour fingerprinting)
   */
  private async simpleHash(str: string): Promise<string> {
    let hash = 0;
    for (let i = 0; i < str.length; i++) {
      const char = str.charCodeAt(i);
      hash = ((hash << 5) - hash) + char;
      hash = hash & hash; // Convert to 32bit integer
    }
    
    // Convertir en string hexadécimal
    return Math.abs(hash).toString(16).padStart(8, '0');
  }

  /**
   * Récupère l'adresse IP de l'utilisateur
   */
  async getUserIpAddress(): Promise<string> {
    if (this.userIpAddress) {
      return this.userIpAddress;
    }

    try {
      // Utiliser un service gratuit pour obtenir l'IP
      const response = await fetch('https://api.ipify.org?format=json');
      const data = await response.json();
      this.userIpAddress = data.ip;
      return data.ip;
    } catch (error) {
      console.error('Error getting IP:', error);
      // Fallback : utiliser une valeur par défaut
      this.userIpAddress = 'unknown';
      return 'unknown';
    }
  }

  /**
   * Vérifie si un email est jetable/temporaire
   */
  isDisposableEmail(email: string): boolean {
    const disposableDomains = [
      'tempmail', 'temp-mail', '10minutemail', 'guerrillamail',
      'mailinator', 'throwaway', 'trashmail', 'fakeinbox',
      'yopmail', 'maildrop', 'getairmail', 'temp-mail.org',
      'sharklasers', 'grr.la', 'guerrillamailblock.com'
    ];

    const domain = email.split('@')[1]?.toLowerCase();
    return disposableDomains.some(d => domain?.includes(d));
  }

  /**
   * Effectue une vérification complète de fraude
   */
  async checkSignupFraud(
    email: string,
    phone?: string,
    siret?: string
  ): Promise<FraudCheckResult> {
    try {
      const deviceFingerprint = await this.generateDeviceFingerprint();
      const ipAddress = await this.getUserIpAddress();

      // Appeler la fonction PostgreSQL
      const { data, error } = await supabase.rpc('check_signup_fraud', {
        p_email: email,
        p_phone: phone || null,
        p_siret: siret || null,
        p_device_fingerprint: deviceFingerprint,
        p_ip_address: ipAddress
      });

      if (error) throw error;

      return {
        fraudScore: data.fraud_score,
        isSuspicious: data.is_suspicious,
        flags: data.flags || [],
        recommendation: data.recommendation
      };
    } catch (error) {
      console.error('Fraud check error:', error);
      // En cas d'erreur, retourner un résultat permissif
      return {
        fraudScore: 0,
        isSuspicious: false,
        flags: [],
        recommendation: 'allow'
      };
    }
  }

  /**
   * Enregistre une tentative d'inscription
   */
  async logSignupAttempt(attempt: SignupAttempt): Promise<void> {
    try {
      const { error } = await supabase
        .from('signup_attempts')
        .insert({
          email: attempt.email,
          phone: attempt.phone,
          device_fingerprint: attempt.deviceFingerprint,
          ip_address: attempt.ipAddress,
          user_agent: attempt.userAgent,
          step_reached: attempt.stepReached,
          success: attempt.success,
          failure_reason: attempt.failureReason
        });

      if (error) throw error;
    } catch (error) {
      console.error('Error logging signup attempt:', error);
      // Ne pas bloquer l'inscription si le logging échoue
    }
  }

  /**
   * Vérifie si un email est disponible
   */
  async checkEmailAvailable(email: string): Promise<boolean> {
    try {
      const { data, error } = await supabase
        .rpc('check_email_available', { p_email: email });

      if (error) throw error;
      return data === true;
    } catch (error) {
      console.error('Error checking email:', error);
      return false;
    }
  }

  /**
   * Vérifie si un SIRET est disponible
   */
  async checkSiretAvailable(siret: string): Promise<boolean> {
    try {
      const cleaned = siret.replace(/[\s]/g, '');
      
      const { data, error } = await supabase
        .rpc('check_siret_available', { p_siret: cleaned });

      if (error) throw error;
      return data === true;
    } catch (error) {
      console.error('Error checking SIRET:', error);
      return false;
    }
  }

  /**
   * Obtient le nombre d'utilisations d'un téléphone
   */
  async getPhoneUsageCount(phone: string): Promise<number> {
    try {
      const cleaned = phone.replace(/[\s.\-]/g, '');
      
      const { data, error } = await supabase
        .rpc('check_phone_usage_count', { p_phone: cleaned });

      if (error) throw error;
      return data || 0;
    } catch (error) {
      console.error('Error checking phone:', error);
      return 0;
    }
  }

  /**
   * Calcule un score de fraude local basique
   */
  async calculateLocalFraudScore(
    email: string,
    phone?: string
  ): Promise<number> {
    let score = 0;

    // Email jetable (+50 points)
    if (this.isDisposableEmail(email)) {
      score += 50;
    }

    // Email suspect (test, fake, etc.)
    if (/test|fake|spam|trash/i.test(email)) {
      score += 30;
    }

    // Téléphone réutilisé
    if (phone) {
      const phoneCount = await this.getPhoneUsageCount(phone);
      if (phoneCount >= 3) {
        score += 40;
      } else if (phoneCount >= 1) {
        score += 20;
      }
    }

    return score;
  }

  /**
   * Réinitialise le cache (pour tests)
   */
  resetCache() {
    this.deviceFingerprint = null;
    this.userIpAddress = null;
  }
}

export const fraudPreventionService = new FraudPreventionService();

File: mobile_flutter/finality_app/lib/services/fraud_prevention_service.dart
import 'package:flutter/material.dart';
import 'package:supabase_flutter/supabase_flutter.dart';
import 'dart:io' show Platform;
import 'package:device_info_plus/device_info_plus.dart';
import '../main.dart';
import '../utils/logger.dart';

/// Service de prévention de fraude lors de l'inscription
/// Détecte les comptes multiples, SIRET dupliqués, emails jetables, etc.
class FraudPreventionService {
  final SupabaseClient _supabase = supabase;

  /// Génère une empreinte unique de l'appareil
  Future<String> getDeviceFingerprint() async {
    final deviceInfo = DeviceInfoPlugin();
    String fingerprint = '';

    try {
      if (Platform.isAndroid) {
        final androidInfo = await deviceInfo.androidInfo;
        fingerprint = '${androidInfo.id}_${androidInfo.model}_${androidInfo.brand}';
      } else if (Platform.isIOS) {
        final iosInfo = await deviceInfo.iosInfo;
        fingerprint = '${iosInfo.identifierForVendor}_${iosInfo.model}_${iosInfo.systemVersion}';
      }
    } catch (e) {
      logger.e('Erreur génération device fingerprint: $e');
      fingerprint = 'unknown_${DateTime.now().millisecondsSinceEpoch}';
    }

    return fingerprint;
  }

  /// Récupère l'adresse IP publique (approximation via geolocation ou API externe)
  /// NOTE: Pour une véritable IP publique, utiliser un service externe comme ipify.org
  Future<String> getPublicIP() async {
    try {
      // TODO: Intégrer un service comme https://api.ipify.org?format=json
      // Pour l'instant, on retourne un placeholder
      return 'unknown_ip';
    } catch (e) {
      logger.e('Erreur récupération IP: $e');
      return 'unknown_ip';
    }
  }

  /// Vérifie si un email est disponible (pas déjà utilisé)
  Future<bool> isEmailAvailable(String email) async {
    try {
      final response = await _supabase.rpc('check_email_available', params: {
        'p_email': email.trim().toLowerCase(),
      });

      return response as bool? ?? false;
    } catch (e) {
      logger.e('Erreur vérification email: $e');
      return false;
    }
  }

  /// Vérifie si un SIRET est disponible (pas déjà utilisé)
  Future<bool> isSiretAvailable(String siret) async {
    try {
      final response = await _supabase.rpc('check_siret_available', params: {
        'p_siret': siret.trim(),
      });

      return response as bool? ?? false;
    } catch (e) {
      logger.e('Erreur vérification SIRET: $e');
      return false;
    }
  }

  /// Compte combien de comptes utilisent ce numéro de téléphone
  Future<int> getPhoneUsageCount(String phone) async {
    try {
      final response = await _supabase.rpc('check_phone_usage_count', params: {
        'p_phone': phone.trim(),
      });

      return response as int? ?? 0;
    } catch (e) {
      logger.e('Erreur vérification téléphone: $e');
      return 0;
    }
  }

  /// Analyse complète de fraude avant inscription
  /// Retourne un objet contenant score, flags, et recommandation
  Future<FraudCheckResult> performFraudCheck({
    required String email,
    String? phone,
    String? siret,
  }) async {
    try {
      final deviceFingerprint = await getDeviceFingerprint();
      final ipAddress = await getPublicIP();

      final response = await _supabase.rpc('check_signup_fraud', params: {
        'p_email': email.trim().toLowerCase(),
        'p_phone': phone?.trim(),
        'p_siret': siret?.trim(),
        'p_device_fingerprint': deviceFingerprint,
        'p_ip_address': ipAddress,
      });

      return FraudCheckResult.fromJson(response as Map<String, dynamic>);
    } catch (e) {
      logger.e('Erreur analyse fraude: $e');
      // En cas d'erreur, on permet l'inscription par défaut (fail-open)
      return FraudCheckResult(
        fraudScore: 0,
        isSuspicious: false,
        flags: [],
        recommendation: FraudRecommendation.allow,
      );
    }
  }

  /// Enregistre une tentative d'inscription (succès ou échec)
  Future<void> logSignupAttempt({
    required String email,
    String? phone,
    required int stepReached,
    required bool success,
    String? failureReason,
  }) async {
    try {
      final deviceFingerprint = await getDeviceFingerprint();
      final ipAddress = await getPublicIP();

      await _supabase.from('signup_attempts').insert({
        'email': email.trim().toLowerCase(),
        'phone': phone?.trim(),
        'device_fingerprint': deviceFingerprint,
        'ip_address': ipAddress,
        'user_agent': Platform.isAndroid ? 'Flutter Android' : 'Flutter iOS',
        'step_reached': stepReached,
        'success': success,
        'failure_reason': failureReason,
      });
    } catch (e) {
      logger.w('Erreur log tentative inscription: $e');
      // Non bloquant
    }
  }

  /// Vérifie des patterns d'email jetables/temporaires
  bool isDisposableEmail(String email) {
    final disposablePatterns = [
      'temp',
      'throw',
      'disposable',
      'trash',
      'fake',
      'test',
      'guerrillamail',
      '10minutemail',
      'mailinator',
      'yopmail',
      'tempmail',
      'trashmail',
    ];

    final lowercase = email.toLowerCase();
    return disposablePatterns.any((pattern) => lowercase.contains(pattern));
  }

  /// Validation format SIRET (14 chiffres)
  bool isValidSiretFormat(String siret) {
    final cleaned = siret.replaceAll(RegExp(r'\s+'), '');
    return RegExp(r'^\d{14}$').hasMatch(cleaned);
  }

  /// Validation format téléphone français (approximatif)
  bool isValidPhoneFormat(String phone) {
    final cleaned = phone.replaceAll(RegExp(r'[\s\.\-\(\)]'), '');
    // Format français: 10 chiffres commençant par 0, ou +33 suivi de 9 chiffres
    return RegExp(r'^(0[1-9]\d{8}|\+33[1-9]\d{8})$').hasMatch(cleaned);
  }

  /// Affiche un message d'avertissement si fraude détectée
  void showFraudWarning(BuildContext context, FraudCheckResult result) {
    if (!result.isSuspicious) return;

    final severity = result.recommendation;
    final message = _getFraudMessage(result);

    showDialog(
      context: context,
      barrierDismissible: severity != FraudRecommendation.block,
      builder: (context) => AlertDialog(
        title: Row(
          children: [
            Icon(
              severity == FraudRecommendation.block
                  ? Icons.block
                  : Icons.warning_amber_rounded,
              color: severity == FraudRecommendation.block
                  ? Colors.red
                  : Colors.orange,
            ),
            const SizedBox(width: 12),
            Text(
              severity == FraudRecommendation.block
                  ? 'Inscription bloquée'
                  : 'Vérification requise',
            ),
          ],
        ),
        content: Text(message),
        actions: [
          if (severity != FraudRecommendation.block)
            TextButton(
              onPressed: () => Navigator.pop(context),
              child: const Text('Continuer quand même'),
            ),
          FilledButton(
            onPressed: () => Navigator.pop(context),
            child: Text(severity == FraudRecommendation.block ? 'OK' : 'Annuler'),
          ),
        ],
      ),
    );
  }

  String _getFraudMessage(FraudCheckResult result) {
    final flags = result.flags.map((f) => f.type).toList();

    if (flags.contains('blacklist_email') ||
        flags.contains('blacklist_phone') ||
        flags.contains('blacklist_siret')) {
      return 'Ces informations sont bloquées. Veuillez contacter le support.';
    }

    if (flags.contains('siret_duplicate')) {
      return 'Ce numéro SIRET est déjà associé à un compte existant.';
    }

    if (flags.contains('device_multiple_accounts')) {
      return 'Plusieurs comptes ont été créés depuis cet appareil. Ceci peut indiquer une activité suspecte.';
    }

    if (flags.contains('phone_multiple_accounts')) {
      return 'Ce numéro de téléphone est utilisé par plusieurs comptes.';
    }

    if (flags.contains('ip_rate_limit')) {
      return 'Trop de tentatives d\'inscription depuis cette connexion. Veuillez réessayer plus tard.';
    }

    if (flags.contains('disposable_email')) {
      return 'Les adresses email temporaires ne sont pas autorisées. Veuillez utiliser un email professionnel.';
    }

    return 'Votre inscription nécessite une vérification manuelle. Vous recevrez un email dans les 24-48h.';
  }
}

/// Résultat d'une vérification anti-fraude
class FraudCheckResult {
  final int fraudScore;
  final bool isSuspicious;
  final List<FraudFlag> flags;
  final FraudRecommendation recommendation;

  FraudCheckResult({
    required this.fraudScore,
    required this.isSuspicious,
    required this.flags,
    required this.recommendation,
  });

  factory FraudCheckResult.fromJson(Map<String, dynamic> json) {
    return FraudCheckResult(
      fraudScore: json['fraud_score'] as int? ?? 0,
      isSuspicious: json['is_suspicious'] as bool? ?? false,
      flags: (json['flags'] as List?)
              ?.map((f) => FraudFlag.fromJson(f as Map<String, dynamic>))
              .toList() ??
          [],
      recommendation: FraudRecommendation.fromString(
        json['recommendation'] as String? ?? 'allow',
      ),
    );
  }

  bool get shouldBlock => recommendation == FraudRecommendation.block;
  bool get needsManualReview => recommendation == FraudRecommendation.manualReview;
}

/// Flag de fraude détecté
class FraudFlag {
  final String type;
  final String severity;

  FraudFlag({required this.type, required this.severity});

  factory FraudFlag.fromJson(Map<String, dynamic> json) {
    return FraudFlag(
      type: json['type'] as String? ?? 'unknown',
      severity: json['severity'] as String? ?? 'low',
    );
  }
}

/// Recommandation suite à analyse de fraude
enum FraudRecommendation {
  allow, // Autoriser l'inscription
  manualReview, // Nécessite vérification manuelle
  block; // Bloquer l'inscription

  static FraudRecommendation fromString(String value) {
    switch (value) {
      case 'block':
        return FraudRecommendation.block;
      case 'manual_review':
        return FraudRecommendation.manualReview;
      default:
        return FraudRecommendation.allow;
    }
  }
}
